feat(HOC): add `withTheme` HOC

kkemple opened the pull request:
<!--
Thanks for your interest in the project. I appreciate bugs filed and PRs submitted!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
Added a HOC (higher order component) called `withTheme` that exposes theme context as a prop to passed in component.

<!-- Why are these changes necessary? -->
**Why**:
It was requested in Issue #41 

<!-- How were these changes implemented? -->
**How**:
I first created a new file to contain the HOC, I then added to global exports. I also added tests.

<!-- feel free to add additional comments -->
There are a few comments in there as some things are unclear still. Mainly, what do we want to do when there is no context found (no ThemeProvider further up the tree). There is also some concern on naming of things. Not sure what you want the exported HOC class name to be.

kkemple commented:
Also, README needs updating still (and do you want an example?). I just wanted to wait until we were good from a code standpoint.

kentcdodds commented on src/with-theme.js:
```diff
@@ -0,0 +1,40 @@
+
+    render() {
+      return (
+        <ComponentToTheme {...Object.assign({}, this.props, this.state)} />
```
This could be simply:

```jsx
<ComponentToTheme {...this.props} {...this.state} />
```

kkemple replied:
:man_facepalming:

kentcdodds commented on src/__tests__/with-theme.js:
```diff
@@ -0,0 +1,71 @@
+  expect(wrapper).toMatchSnapshotWithGlamor(`with theme prop of padding 20px`)
+})
+
+test('acts as pass through when no theme provider found up tree', () => {
```
Do we have any way of detecting whether there's a provider up the tree? If so, maybe we should log a warning?

kkemple replied:
i had a console.warn statement in at first but it broke linting (not that it can't be ignored). I don't think you can really tell if you have a provider, but you can definitely warn if there is no `__glamorous__` context prop (which is a good indicator of that situation).

kentcdodds replied:
I think let's just leave it as it is for now... Maybe we'll add a warning in the future. Thanks!

kkemple replied:
oops just saw this, will revert that lol

kentcdodds commented on src/__tests__/with-theme.js:
```diff
@@ -0,0 +1,71 @@
+/* eslint func-style:0 */
+import React, {Component} from 'react'
+import {render, mount} from 'enzyme'
+import * as jestGlamorReact from 'jest-glamor-react'
```
Because you're using the `style` prop to apply styles (which I think is actually really smart), you shouldn't need this. You can just use `toMatchSnapshot` :)

kkemple replied:
yeah i mean to update that :blush:

kkemple commented:
Updated per your review comments, let me know if there is anything I missed. I added a `console.warn` statement if there is no `__glamorous__` context found. So be sure to give that a review as well!

kkemple commented:
removed `console.warn` statement if no context found 😄

kkemple commented:
👍

kwelch commented:
Do you think there should be a test that ensures it works when wrapping a class?

Other than that, it looks good.

kentcdodds commented:
> Also, do you want any info about this new API in README?

Oh! Of course! Thanks for mentioning it. Yes, let's please have a section about this right under the docs about the `ThemeProvider` :+1:

vesparny commented on src/with-theme.js:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    componentDidMount() {
+      if (this.context[CHANNEL]) {
```
`withTheme()` is supposed to be used inside a `<ThemeProvider />`.
Shouldn't we throw an error here if `!this.context[CHANNEL]` ?

This would warn the user before he realize he does not have access to `theme` in children glamorous comps.

kentcdodds replied:
We actually [discussed](https://github.com/paypal/glamorous/pull/45#discussion_r111036984) adding a warning (and @kkemple even implemented it before I said we should just leave it). Maybe we should add it back after all. I would want it to be wrapped in a `process.env.NODE_ENV === 'production'` call though. I'm uncertain whether our rollup config is currently capable of removing this for our UMD build. May need to investigate that a bit.

kkemple replied:
I added this back in under a process.env if statement so it should be tree shakeable once the build part is figured back out. I had to ignore the console statement for eslint and the if statement for instanbul since there is no way to test that statement safely (that I am aware of).

kkemple commented:
I've added a section to the README and implemented a warning in production when there is no `ThemeProvider` further up the tree. :smile:

kentcdodds commented on README.md:
````diff
@@ -344,6 +344,72 @@ const Title = glamorous.h1({
 ```
 
+`glamorous` also exports a `withTheme` higher order component (HOC) so you can access your theme in any component!
+> Try this out in your browser [here](https://codesandbox.io/s/qYmJjE4jy)!
````
Very cool! When this gets merged and released, you'll need to update the glamorous version in your packages on that codesandbox :)

kkemple replied:
oh right, good call!

kentcdodds commented on README.md:
````diff
@@ -344,6 +344,72 @@ const Title = glamorous.h1({
+</ThemeProvider>
+```
+
+Or if you prefer decorator syntax:
````
Didn't realize it worked like this out of the box. Neat!

kkemple replied:
Yeah I haven't looked back since I discovered that! :smile:
I have great one that is for checking internet connection `@connection` adds an offline prop that is a boolean. It's great.

kentcdodds commented on README.md:
````diff
@@ -344,6 +344,72 @@ const Title = glamorous.h1({
+  <SubTitle>from withTheme!</SubTitle>
+</ThemeProvider>
+```
+> `withTheme` expects a `ThemeProvider` further up the render tree and will warn in `production` if one is not found!
````
Should only warn in `development` :+1:

kkemple replied:
oh oops, i thought you said in production, will update :smile:

kentcdodds commented on src/with-theme.js:
```diff
@@ -0,0 +1,46 @@
+    componentWillMount() {
+      if (!this.context[CHANNEL]) {
+        /* istanbul ignore if */
+        if (process.env.NODE_ENV === 'production') {
```
This should be: `if (process.env.NODE_ENV !== 'production') {`

We don't want to warn in production :)

kentcdodds commented on src/with-theme.js:
```diff
@@ -0,0 +1,46 @@
+      if (!this.context[CHANNEL]) {
+        /* istanbul ignore if */
+        if (process.env.NODE_ENV === 'production') {
+          /* eslint-disable no-console */
```
Could we use this instead:

```
// eslint-disable-next-line no-console
```

kkemple replied:
yeah we can for sure

kkemple replied:
did not know about `eslint-disable-next-line` much better!

kentcdodds commented on src/__tests__/with-theme.js:
```diff
@@ -0,0 +1,101 @@
+  const Child = withTheme(() => <div />)
+  const Parent = () => <Child />
+  const wrapper = mount(<Parent />)
+  expect(wrapper).toMatchSnapshot(`without theme provider`)
```
Could we add an assertion here that `console.warn` was called? We'd need to mock `console.warn`:

```javascript
const originalWarn = console.warn
console.warn = jest.fn()
// your test code you have already
expect(console.warn).toHaveBeenCalledTimes(1)
expect(console.warn). toHaveBeenCalledWith('Expected ThemeProvider but none found!')
```

And actually, if we could make that warning a little more helpful that'd be great. Like providing `ComponentToTheme .displayName || ComponentToTheme .name` and saying it's coming from glamorous. So like:

```javascript
// eslint-ignore-next-line max-len
console.warn(`glamorous warning: Expected component called ${ComponentToTheme .displayName || ComponentToTheme .name} which uses withTheme to be within a ThemeProvider but none was found.`)
```

I don't mind being verbose about this :smile:

kkemple replied:
I figured there would be some adjustment there, just filled in anything until you figured out warning message.

kkemple commented:
@kentcdodds made requested changes!

kentcdodds commented on src/with-theme.js:
```diff
@@ -0,0 +1,57 @@
+
+function generateWarningMessage(componentName) {
+  // eslint-disable-next-line max-len
+  return `glamorous warning: Expected component called ${componentName} which uses withTheme to be within a ThemeProvider but none was found.`
```
Why don't we wrap `componentName` this in quotes?

kkemple replied:
done!

kentcdodds commented on src/with-theme.js:
```diff
@@ -0,0 +1,57 @@
+
+    componentWillMount() {
+      if (!this.context[CHANNEL]) {
+        // istanbul ignore else
```
Actually, I'd really like to make sure that we don't log in production. So let's add a test that basically does:

```javascript
const originalEnv = process.env.NODE_ENV
process.env.NODE_ENV = 'production'

// tests

process.env.NODE_ENV = originalEnv
```

Thanks!

kkemple replied:
added!

kentcdodds commented on src/__tests__/with-theme.js:
```diff
@@ -0,0 +1,130 @@
+  const Parent = () => <Child />
+  mount(<Parent />)
+
+  expect(console.warn).toHaveNotBeenCalled
```
This won't work I'm afraid. Should be:

```js
expect(console.warn).not.toHaveBeenCalled()
```

kkemple replied:
oh right! lol

kkemple commented:
okay, updated!

kentcdodds commented:
@kkemple, thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the CONTRIBUTING.md file (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

kkemple commented:
Thanks! I look forward to contributing and helping others do the same! :smile:
